### PR TITLE
add timestamp to global state table for lease table cleanup

### PR DESCRIPTION
**Pull request by MaddyDev**
Added LastAccessTime Datetime column to the GlobalState table. For accommodating existing triggers, added the column with default getutcdate() value.

Added the clean up scripts to the BindingsOveriew doc, I could move them over to a more appropriate place if folks have better suggestions.

### CODE REVIEW

File: docs/BindingsOverview.md [hunk 1]
```diff
@@ -168,3 +169,80 @@ e.g.
 ```sql
 UPDATE [Products].[az_func].[Leases_<FunctionId>_<TableId>] SET _az_func_AttemptCount = 0 WHERE ProductId = 123
 ```
+
+#### Lease Tables clean up
+
+Before clean up, please see [Leases table](./TriggerBinding.md#az_funcleases_) documentation for how to get the correct Leases table.
+
+- Clean up all the tables not accessed more than N number of days, Please initialize N according to your use case:
```
Comment (Charles-Gagnon) [inline docs/BindingsOverview.md:177]:
```suggestion
- Deletes all the lease tables that haven't been accessed in N days:
```

File: docs/BindingsOverview.md [hunk 2]
```diff
@@ -168,3 +169,80 @@ e.g.
...
+
+```sql
+DECLARE @table_name NVARCHAR(MAX);
```
Comment (Charles-Gagnon) [inline docs/BindingsOverview.md:180]: Add a `USE [<DbName>]` at the top with instructions to set it to the database to be cleaned up

File: docs/BindingsOverview.md [hunk 3]
```diff
@@ -168,3 +169,80 @@ e.g.
...
+DECLARE @UserFunctionId char(16);
+DECLARE @UserTableId int;
+DECLARE @N int;
```
Comment (Charles-Gagnon) [inline docs/BindingsOverview.md:183]: I would also suggest naming this `CleanupAgeDays` or something similar to make it clearer

```suggestion
DECLARE @N int = <Insert desired cleanup age in days here>`
```

File: docs/BindingsOverview.md [hunk 4]
```diff
@@ -168,3 +169,80 @@ e.g.
...
+
+#### Lease Tables clean up
+
+Before clean up, please see [Leases table](./TriggerBinding.md#az_funcleases_) documentation for how to get the correct Leases table.
+
+- Clean up all the tables not accessed more than N number of days, Please initialize N according to your use case:
+
+```sql
+DECLARE @table_name NVARCHAR(MAX);
+DECLARE @UserFunctionId char(16);
```
Comment (Charles-Gagnon) [inline docs/BindingsOverview.md:181]: Please also add a comment explaining what the user should set this value to and how to get it so the script itself has context about what it's doing.

```suggestion
DECLARE @UserFunctionId char(16) = <Insert function ID here>`
```

File: docs/BindingsOverview.md (hunk 2 above)
Comment (Charles-Gagnon) [inline docs/BindingsOverview.md:180]: Would suggest putting a comment at the top explaining what the script is doing as well so people can simply copy and paste it and not have to add their own context in (for example if they make it into a job)

File: docs/BindingsOverview.md [hunk 5]
```diff
@@ -168,3 +169,80 @@ e.g.
...
+DECLARE @UserTableId int;
+DECLARE @N int;
+DECLARE leasetable_cursor CURSOR FOR
```
Comment (Charles-Gagnon) [inline docs/BindingsOverview.md:184]: nit: LeaseTable_Cursor (keep consistent capitalization in the script)

File: docs/BindingsOverview.md [hunk 6]
```diff
@@ -168,3 +169,80 @@ e.g.
...
+- Clean up a specific lease table concerning a specific function:
+
+To find the name of the leases table associated with your function, look in the log output for a line such as this which is emitted when the trigger is started.
+
+`SQL trigger Leases table: [az_func].[Leases_84d975fca0f7441a_901578250]`
+
+This log message is at the `Information` level, so make sure your log level is set correctly.
+
+```sql
+DECLARE @table_name NVARCHAR(MAX); --@table_name would be [az_func].[Leases_84d975fca0f7441a_901578250]
```
Comment (Charles-Gagnon) [inline docs/BindingsOverview.md:215]: Same here about doing `DECLARE @... = <Insert value here>` type stuff so the user can't run it until they fill in those values. 

You may also want to change the comments you have here to explain what the values are and then say something like `, e.g. '[az_func].[Leases_84d975fca0f7441a_901578250]'` to indicate that these are sample values of what they should look like

File: docs/BindingsOverview.md [hunk 7]
```diff
@@ -168,3 +169,80 @@ e.g.
...
+DECLARE @UserFunctionId char(16); -- @UserFunctionId would be 84d975fca0f7441a
+DECLARE @UserTableId int; -- @UserTableId would be 901578250
+
+DROP TABLE IF EXISTS @table_name
+DELETE FROM az_func.GlobalState WHERE UserFunctionID = @UserFunctionId and UserTableID = @UserTableId
+```
+
+- Clear all trigger related data for a reset:
```
Comment (Charles-Gagnon) [inline docs/BindingsOverview.md:223]: Maybe just have the first script be able to take in -1 for the days value to clean up all tables instead? Help keep things simpler

File: docs/BindingsOverview.md (hunk 6 above)
Comment (Charles-Gagnon) [inline docs/BindingsOverview.md:215]: nit: `@TableName` to keep casing consistent

File: docs/BindingsOverview.md [hunk 8]
```diff
@@ -168,3 +169,80 @@ e.g.
 ```sql
 UPDATE [Products].[az_func].[Leases_<FunctionId>_<TableId>] SET _az_func_AttemptCount = 0 WHERE ProductId = 123
 ```
+
+#### Lease Tables clean up
+
+Before clean up, please see [Leases table](./TriggerBinding.md#az_funcleases_) documentation for how to get the correct Leases table.
```
Comment (Charles-Gagnon) [inline docs/BindingsOverview.md:175]: Add comment here explaining why cleanup is necessary and when the user would want to do it.

File: docs/BindingsOverview.md [hunk 9]
```diff
@@ -168,3 +169,93 @@ e.g.
...
+    DELETE FROM az_func.GlobalState WHERE UserFunctionID = @UserFunctionId and UserTableID = @UserTableId
+    FETCH NEXT FROM LeaseTable_Cursor INTO @TableName, @UserFunctionId, @UserTableId;
+END;
+
+CLOSE LeaseTable_Cursor;
+
+DEALLOCATE LeaseTable_Cursor;
+```
+
+- Clean up a specific lease table concerning a specific function:
```
Comment (Charles-Gagnon) [inline docs/BindingsOverview.md:215]:
```suggestion
- Clean up a specific lease table:
```

File: docs/BindingsOverview.md [hunk 10]
```diff
@@ -168,3 +169,93 @@ e.g.
...
+
+To find the name of the leases table associated with your function, look in the log output for a line such as this which is emitted when the trigger is started.
```
Comment (Charles-Gagnon) [inline docs/BindingsOverview.md:217]:
```suggestion
To find the name of the lease table associated with your function, look in the log output for a line such as this which is emitted when the trigger is started.
```

File: docs/BindingsOverview.md [hunk 11]
```diff
@@ -168,3 +169,93 @@ e.g.
...
+
+`SQL trigger Leases table: [az_func].[Leases_84d975fca0f7441a_901578250]`
+
+This log message is at the `Information` level, so make sure your log level is set correctly.
+
+```sql
+-- Deletes the specified lease table and removes it from GlobalState table.
+USE <Insert DATABASE name here>;
+DECLARE @TableName NVARCHAR(MAX) = <Insert lease table name here>; -- e.g. '[az_func].[Leases_84d975fca0f7441a_901578250]
+DECLARE @UserFunctionId char(16) = <Insert function ID here>; -- e.g. '84d975fca0f7441a'
```
Comment (Charles-Gagnon) [inline docs/BindingsOverview.md:227]: Add comments explaining how they get these values (from the leases table name)

File: docs/BindingsOverview.md [hunk 12]
```diff
@@ -168,3 +169,93 @@ e.g.
...
+
+#### Lease Tables clean up
+
+Before clean up, please see [Leases table](./TriggerBinding.md#az_funcleases_) documentation for understanding how they are created and used.
+
+Why clean up?
+1. You renamed your function/class/method name, which causes a new lease table to be created and the old one to be obsolete.
+2. You created a trigger function that you no longer need and wish to clean up its associated data.
+3. You want to reset your environment.
+The Azure SQL Trigger does not currently handle cleaning up any leftover objects, and so to assist you with that we have provided the below scripts to help guide you through doing that.
```
Comment (Charles-Gagnon) [inline docs/BindingsOverview.md:181]:
```suggestion

The Azure SQL Trigger does not currently handle automatically cleaning up any leftover objects, and so we have provided the below scripts to help guide you through doing that.
```

File: src/TriggerBinding/SqlTriggerListener.cs [hunk 13]
```diff
@@ -321,8 +321,12 @@ private async Task<long> CreateGlobalStateTableAsync(SqlConnection connection, S
                         UserFunctionID char(16) NOT NULL,
                         UserTableID int NOT NULL,
                         LastSyncVersion bigint NOT NULL,
+                        LastAccessTime Datetime NULL,
```
Comment (Charles-Gagnon) [inline src/TriggerBinding/SqlTriggerListener.cs:324]: Let's change this to NOT NULL and default to GETDATE. That makes it simpler, and there's no reason I can think of to have this actually be NULL

File: src/TriggerBinding/SqlTriggerListener.cs [hunk 14]
```diff
@@ -321,8 +321,12 @@ private async Task<long> CreateGlobalStateTableAsync(SqlConnection connection, S
...
                         PRIMARY KEY (UserFunctionID, UserTableID)
                     );
+                ELSE IF NOT EXISTS(SELECT 1 FROM sys.columns WHERE Name = N'LastAccessTime'
```
Comment (chlafreniere) [inline src/TriggerBinding/SqlTriggerListener.cs:327]: I would like for us to have an explicit test for a lease table where the LastAccessTime column doesn't exist to validate that the `ELSE IF NOT EXISTS` code path works as intended.

(Is this possible, by the way, in the real world?)

File: src/TriggerBinding/SqlTableChangeMonitor.cs [hunk 15]
```diff
@@ -955,7 +955,7 @@ private SqlCommand BuildUpdateTablesPostInvocation(SqlConnection connection, Sql
                 IF @unprocessed_changes = 0 AND @current_last_sync_version < {newLastSyncVersion}
                 BEGIN
                     UPDATE {GlobalStateTableName}
-                    SET LastSyncVersion = {newLastSyncVersion}
+                    SET LastSyncVersion = {newLastSyncVersion}, LastAccessTime = GETDATE()
```
Comment (chlafreniere) [inline src/TriggerBinding/SqlTableChangeMonitor.cs:958]: Should we be using GETDATE() or GETUTCDATE()?

File: docs/BindingsOverview.md [hunk 16]
```diff
@@ -168,3 +169,93 @@ e.g.
...
+FROM az_func.GlobalState
+WHERE DATEDIFF(day, LastAccessTime, GETDATE()) > @CleanupAgeDays
+
+OPEN LeaseTable_Cursor;
+
+FETCH NEXT FROM LeaseTable_Cursor INTO @TableName, @UserFunctionId, @UserTableId;
+
+WHILE @@FETCH_STATUS = 0
+BEGIN
+    EXEC ('DROP TABLE IF EXISTS ' + @TableName);
```
Comment (Charles-Gagnon) [inline docs/BindingsOverview.md:205]: Let's add print statements to these, put the appropriate one above each statement that drops/deletes (here and below)

`PRINT 'Dropping table ' + @TableName`
`PRINT 'Removing row from GlobalState for UserFunctionID = ' + @UserFunctionId + ' and UserTableID = ' + @UserTableId'`

File: docs/BindingsOverview.md [hunk 17]
```diff
@@ -168,3 +169,93 @@ e.g.
...
+This log message is at the `Information` level, so make sure your log level is set correctly.
+
+```sql
+-- Deletes the specified lease table and removes it from GlobalState table.
+USE <Insert DATABASE name here>;
+DECLARE @TableName NVARCHAR(MAX) = <Insert lease table name here>; -- e.g. '[az_func].[Leases_84d975fca0f7441a_901578250]
+DECLARE @UserFunctionId char(16) = <Insert function ID here>; -- e.g. '84d975fca0f7441a' the first section of the lease table name [Leases_84d975fca0f7441a_901578250].
+DECLARE @UserTableId int = <Insert table ID here>; -- e.g. '901578250' the second section of the lease table name [Leases_84d975fca0f7441a_901578250].
+
+DROP TABLE IF EXISTS @TableName
```
Comment (Charles-Gagnon) [inline docs/BindingsOverview.md:230]:
```suggestion
EXEC ('DROP TABLE IF EXISTS ' + @TableName);
```

**Review by Charles-Gagnon — COMMENTED** [review]
Make sure you run through and test all the scripts - there was one bug I found doing it myself. Didn't see any others, but it'd be best to double check yourself.

File: src/TriggerBinding/SqlTriggerListener.cs (hunk 14 above)
Comment (Charles-Gagnon) [reply src/TriggerBinding/SqlTriggerListener.cs:327]: It will be for people with existing functions that just update the extension version but everything else stays the same. But that'll drop off as we march towards GA since after this change gets in no one should have that going forward.

File: src/TriggerBinding/SqlTableChangeMonitor.cs (hunk 15 above)
Comment (Charles-Gagnon) [reply src/TriggerBinding/SqlTableChangeMonitor.cs:958]: It shouldn't matter I think - it's all local to the server (so unless that changes timezones or something weird). 

That being said - using UTC seems reasonable to avoid that potential edge case. And that's a good general thing to default to for stored timestamps anyways.

Comment (MaddyDev) [reply src/TriggerBinding/SqlTableChangeMonitor.cs]: Defaulted to GETUTCDATE() to avoid potential edge cases.

File: src/TriggerBinding/SqlTriggerListener.cs (hunk 14 above)
Comment (MaddyDev) [reply src/TriggerBinding/SqlTriggerListener.cs:327]: I agree with @Charles here, but I can look into adding a test to be sure nothing breaks in the meantime.

**Review by Charles-Gagnon — APPROVED** [review]
I agree a test would be good, but the rest looks fine other than that.

**Comment by MaddyDev** [comment]
Merging this in, will send a separate PR for the test.
